common/component/postgresql/awsIAM/aws.go

```diff
@@ -0,0 +1,240 @@
+	dbEndpoint := fmt.Sprintf("%s:%d", pgCfg.Host, pgCfg.Port)
+	var authenticationToken string
+
+	// https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/UsingWithRDS.IAMDBAuth.Connecting.Go.html
+	// Default to load default config through aws credentials file (~/.aws/credentials)
+	awsCfg, err := aws_config.LoadDefaultConfig(ctx)
+	if err != nil {
+		// otherwise use metadata fields
+
+		// Validate if access key and secret access key are provided
+		if accessKey == "" || secretKey == "" {
+			return "", fmt.Errorf("failed to load default configuration for AWS using accessKey and secretKey")
+		}
+
+		// Set credentials explicitly
+		var awsCfg2 aws.CredentialsProvider
+		awsCfg2 = aws_credentials.NewStaticCredentialsProvider(accessKey, secretKey, "")
+		authenticationToken, err = aws_auth.BuildAuthToken(
+			ctx, dbEndpoint, region, pgCfg.User, awsCfg2)
+		if err != nil {
+			return "", fmt.Errorf("failed to create AWS authentication token: %v", err)
+		}
+
+		return authenticationToken, nil
+	}
+
+	authenticationToken, err = aws_auth.BuildAuthToken(
+		ctx, dbEndpoint, region, pgCfg.User, awsCfg.Credentials)
+	if err != nil {
+		return "", fmt.Errorf("failed to create AWS authentication token: %v", err)
+	}
+
+	return authenticationToken, nil
+}
+
+func parseDatabaseFromConnectionString(connectionString string) (string, error) {
+	match := databaseNameRegex.FindStringSubmatch(connectionString)
+	if len(match) < 2 {
+		return "", fmt.Errorf("unable to find database name ('dbname' field) in the connection string")
+	}
+
+	// Return the value after "dbname="
+	return match[1], nil
+}
+
+func parseUserFromConnectionString(connectionString string) (string, error) {
+	match := databaseUserRegex.FindStringSubmatch(connectionString)
+	if len(match) < 2 {
+		return "", fmt.Errorf("unable to find database user in the connection string")
+	}
+
+	// Return the value after "user="
+	return match[1], nil
+}
+
+func parsePasswordFromConnectionString(connectionString string) (string, error) {
+	match := databasePasswordRegex.FindStringSubmatch(connectionString)
+	if len(match) < 2 {
+		return "", fmt.Errorf("unable to find database password field in the connection string")
+	}
+
+	// Return the value after "password="
+	return match[1], nil
+}
+
+func CreateDatabaseIfNeeded(ctx context.Context, timeout time.Duration, connectionString string, db pginterfaces.PGXPoolConn) error {
+	// Parse database name from connection string
+	dbName, err := parseDatabaseFromConnectionString(connectionString)
+	if err != nil {
+		return fmt.Errorf("failed to parse database name from connection string to create database %v", err)
+	}
+
+	// check if database exists in connection string
+	var dbExists bool
+	dbExistsCtx, dbExistsCancel := context.WithTimeout(ctx, timeout)
+	err = db.QueryRow(dbExistsCtx, databaseExistsTmpl, dbName).Scan(&dbExists)
+	dbExistsCancel()
+	if err != nil && err != pgx.ErrNoRows {
```

Comment:
Should this be an OR?

```suggestion
	if err != nil || err != pgx.ErrNoRows {
```

Reply:
Actually, maybe you are right. Maybe it makes sense to err only if there is an err and the err is not no rows found. If the err is no rows found we probably shouldn't error out since the query just returned empty, like no rows. I'll dismiss my comment here.